pgx/src/datum/range.rs

```diff
@@ -0,0 +1,371 @@
+Portions Copyright 2021-2022 Technology Concepts & Design, Inc. <support@tcdi.com>
+
+All rights reserved.
+
+Use of this source code is governed by the MIT license that can be found in the LICENSE file.
+*/
+
+//! Utility functions for working with `pg_sys::RangeType` structs
+use crate::{
+    pg_sys, void_mut_ptr, Date, FromDatum, IntoDatum, Numeric, Timestamp, TimestampWithTimeZone,
+};
+use pgx_pg_sys::{Oid, RangeBound};
+use pgx_utils::sql_entity_graph::metadata::{
+    ArgumentError, Returns, ReturnsError, SqlMapping, SqlTranslatable,
+};
+use std::marker::PhantomData;
+
+/// Represents Datum to serialized RangeType PG struct
+pub struct Range<T: FromDatum + IntoDatum + RangeSubType> {
+    ptr: *mut pg_sys::varlena,
+    range_type: *mut pg_sys::RangeType,
+    _marker: PhantomData<T>,
+}
+
+impl<T> Range<T>
+where
+    T: FromDatum + IntoDatum + RangeSubType,
+{
+    /// ## Safety
+    /// function requires that
+    /// - datum is null OR
+    /// - datum represents a RangeType datum
+    #[inline]
+    unsafe fn from_pg(ptr: *mut pg_sys::varlena, range_type: *mut pg_sys::RangeType) -> Self {
+        Range { ptr, range_type, _marker: PhantomData }
+    }
+}
+
+impl<T> TryFrom<pg_sys::Datum> for Range<T>
+where
+    T: FromDatum + IntoDatum + RangeSubType,
+{
+    type Error = RangeConversionError;
+
+    /// ## Safety
+    /// function requires that
+    /// - datum is null OR
+    /// - datum represents a RangeType datum
+    fn try_from(datum: pg_sys::Datum) -> Result<Self, Self::Error> {
+        if datum.is_null() {
+            Err(RangeConversionError::NullDatum)
+        } else {
+            unsafe {
+                let ptr = datum.cast_mut_ptr();
+                let range_type =
+                    pg_sys::pg_detoast_datum(datum.cast_mut_ptr()) as *mut pg_sys::RangeType;
+
+                Ok(Range::<T>::from_pg(ptr, range_type))
+            }
+        }
+    }
+}
+
+impl<T> FromDatum for Range<T>
+where
+    T: FromDatum + IntoDatum + RangeSubType,
+{
+    #[inline]
+    unsafe fn from_polymorphic_datum(
+        datum: pg_sys::Datum,
+        is_null: bool,
+        _: pg_sys::Oid,
+    ) -> Option<Self>
+    where
+        Self: Sized,
+    {
+        if is_null {
+            None
+        } else {
+            Some(datum.try_into().expect("Error converting RangeType datum"))
+        }
+    }
+}
+
+impl<T> IntoDatum for Range<T>
+where
+    T: FromDatum + IntoDatum + RangeSubType,
+{
+    #[inline]
+    fn into_datum(self) -> Option<pg_sys::Datum> {
+        Some(self.range_type.into())
+    }
+
+    #[inline]
+    fn type_oid() -> pg_sys::Oid {
+        T::range_type_oid()
+    }
+}
+
+impl<T> Drop for Range<T>
+where
+    T: FromDatum + IntoDatum + RangeSubType,
+{
+    fn drop(&mut self) {
+        if !self.range_type.is_null() && self.range_type as *mut pg_sys::varlena != self.ptr {
+            unsafe {
+                pg_sys::pfree(self.range_type as void_mut_ptr);
+            }
+        }
```

Comment:
It would be a good idea to document why we need this unsafe block and why those preconditions have to be upheld.

Reply:
Another example of me blindly copying [0.4.5 array.rs logic](https://github.com/tcdi/pgx/blob/1a89778046547c52c714fe6885069eab5638f965/pgx/src/datum/array.rs#L266
) when working with a varlena/detoasted type. 
I'm still too green at Rust to tell you what's right or wrong here.  added safety messaging.